ibis/backends/flink/__init__.py

```diff
@@ -635,6 +638,111 @@ def drop_view(
         sql = statement.compile()
         self._exec_sql(sql)
 
+    def _get_dataframe_from_path(self, path: str | Path) -> pd.DataFrame:
+        import glob
+
+        import pandas as pd
+
+        dataframe_list = []
+        path_str = str(path)
+        path_normalized = normalize_filename(path)
+        for file_path in glob.glob(path_normalized):
+            if path_str.startswith(("parquet://", "parq://")) or path_str.endswith(
+                ("parq", "parquet")
+            ):
+                dataframe = pd.read_parquet(file_path)
+            elif path_str.startswith("csv://") or path_str.endswith(("csv", "csv.gz")):
+                dataframe = pd.read_csv(file_path)
+            elif path_str.endswith("json"):
+                dataframe = pd.read_json(file_path, lines=True)
+            else:
+                raise ValueError(f"Unsupported file_path: {file_path}")
+
+            dataframe_list.append(dataframe)
+
+        return pd.concat(dataframe_list, ignore_index=True, sort=False)
+
+    def read_file(
+        self,
+        file_type: str,
+        path: str | Path,
+        table_name: str | None = None,
+    ) -> ir.Table:
+        """Register a file as a table in the current database.
+
+        Parameters
+        ----------
+        file_type
+            File type, e.g., parquet, csv, json.
+        path
+            The data source.
+        table_name
+            An optional name to use for the created table. This defaults to
+            a sequentially generated name.
+
+        Returns
+        -------
+        ir.Table
+            The just-registered table
+        """
+        obj = self._get_dataframe_from_path(path)
```

Comment:
Does flink support natively loading directly from files? If so, I think we should use that.

If it doesn't, I'm not sure how I feel about automatically loading files using pandas and forwarding them that way. We do something similar for local in-memory backends like duckdb/pandas where file paths are unambiguously local. But for a potentially distributed system like flink, automatically using a local file reader may have unexpected behavior with path names, and also may be inefficient.

If we do decide to handle file reading manually using an in-memory reader here, we should use the ones provided by `pyarrow` directly, not pandas. These have builtin support for directory datasets, and better match the behavior of other backends.

Reply:
Thanks, I agree that reading the file is not ideal while creating the table. Flink actually supports creating tables with [filesystem connector](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/filesystem/). This however requires specifying the schema, as `create_table()` also requires:
```python
def create_table(
        self,
        name: str,
        obj: pd.DataFrame | pa.Table | ir.Table | None = None,
        *,
        schema: sch.Schema | None = None,
        database: str | None = None,
        catalog: str | None = None,
        tbl_properties: dict | None = None,
        watermark: Watermark | None = None,
        temp: bool = False,
        overwrite: bool = False,
    ) -> ir.Table:
        ...

               if obj is None and schema is None:
                   raise exc.IbisError("`schema` or `obj` is required")
    ...
```

So I think we have two options:
1. Add a required argument `schema` for `read_***()`. This would deviate from the interface existing for other backends.
2. "Read" the schema from the file with `pyarrow` and feed it into `create_table()` with `filesystem` connector. This has the same issue you raised in accessing the files in a distributed system.

Which one do you think makes more sense? We could also implement both where the user can specify the schema, and if not specified we could construct the schema from the file and raise an error in case of a failure in accessing the file.